src/Microsoft.AspNet.OData.Shared/Builder/LinkGenerationHelpers.cs

```diff
@@ -550,5 +558,54 @@ internal static IList<ODataPathSegment> GenerateBaseODataPathSegments(this Resou
                 feedContext.EntitySetBase,
                 odataPath);
         }
+
+        private static IList<ODataPathSegment> GenerateContainmentODataPathSegments(this ResourceContext resourceContext)
+        {
+            IList<ODataPathSegment> navigationPathSegments = new List<ODataPathSegment>();
+            ResourceContext currentResourceContext = resourceContext;
+
+            while (currentResourceContext != null)
+            {
+                IEdmNavigationSource navigationSource = currentResourceContext.NavigationSource;
+
+                if (navigationSource is IEdmContainedEntitySet containedEntitySet)
+                {
+                    KeySegment keySegment = new KeySegment(
+                        ConventionsHelpers.GetEntityKey(currentResourceContext),
+                        currentResourceContext.StructuredType as IEdmEntityType,
+                        navigationSource: null);
+                    navigationPathSegments.Insert(0, keySegment);
+
+                    NavigationPropertySegment navPropertySegment = new NavigationPropertySegment(
+                        containedEntitySet.NavigationProperty,
+                        containedEntitySet.ParentNavigationSource);
+                    navigationPathSegments.Insert(0, navPropertySegment);
+                }
+                else if (navigationSource is IEdmEntitySet entitySet)
+                {
+                    KeySegment keySegment = new KeySegment(
+                        ConventionsHelpers.GetEntityKey(currentResourceContext),
+                        currentResourceContext.StructuredType as IEdmEntityType,
+                        null);
+                    navigationPathSegments.Insert(0, keySegment);
+
+                    EntitySetSegment entitySetSegment = new EntitySetSegment(entitySet);
+                    navigationPathSegments.Insert(0, entitySetSegment);
+
+                    break;
+                }
+                else if (navigationSource is IEdmSingleton singleton)
+                {
+                    SingletonSegment singletonSegment = new SingletonSegment(singleton);
+                    navigationPathSegments.Insert(0, singletonSegment);
+
+                    break;
+                }
+
+                currentResourceContext = currentResourceContext.SerializerContext.ExpandedResource;
+            }
+
+            return navigationPathSegments;
```

Comment:
How do we handle canonical urls?

Reply:
Yes we do. You can take a look at the test `VerifyExpectedNextLinksGeneratedForNonContainedNavigationPropertyDeclaredOnDerivedTypeAsODataPathSegment` to confirm